pkg/tests/testsupport/metric_loader.go

```diff
@@ -0,0 +1,245 @@
+		"series", it.blocks[it.curBlockIdx].Meta().Stats.NumSeries)
+	it.blockSamples = make([]*BlockSample, it.blocks[it.curBlockIdx].Meta().Stats.NumSamples)
+	it.labelsCache = make(map[uint64]labels.Labels, it.blocks[it.curBlockIdx].Meta().Stats.NumSeries)
+	querier, err := tsdb.NewBlockQuerier(it.blocks[it.curBlockIdx], math.MinInt64, math.MaxInt64)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := querier.Close(); err != nil {
+			log.Error(err)
+		}
+	}()
+	seriesSet := querier.Select(false, nil, allMatcher)
+	sampleCounter := 0
+	for seriesSet.Next() {
+		series := seriesSet.At()
+		seriesIt := series.Iterator()
+		lblsHash := series.Labels().Hash()
+		if _, ok := it.labelsCache[lblsHash]; !ok {
+			it.labelsCache[lblsHash] = series.Labels()
+		}
+		for seriesIt.Next() {
+			ts, val := seriesIt.At()
+			sample := &BlockSample{ts, val, lblsHash}
+			it.blockSamples[sampleCounter] = sample
+			sampleCounter++
+		}
+	}
+	sort.SliceStable(it.blockSamples, func(i, j int) bool {
+		return it.blockSamples[i].ts < it.blockSamples[j].ts
+	})
+	return nil
+}
+
+// Get returns sample at current position
+func (i *promIterator) Get() timeSeries {
+	blockSample := i.blockSamples[i.curSampleIdx]
+	labels := i.labelsCache[blockSample.lblsHash]
+	protoLabels := make([]prompb.Label, len(labels))
+	for i, l := range labels {
+		protoLabels[i] = prompb.Label{Name: l.Name, Value: l.Value}
+	}
+	sample := prompb.Sample{
+		Value:     blockSample.val,
+		Timestamp: blockSample.ts,
+	}
+	ts := prompb.TimeSeries{
+		Labels:               protoLabels,
+		Samples:              []prompb.Sample{sample},
+		Exemplars:            []prompb.Exemplar{},
+		XXX_NoUnkeyedLiteral: struct{}{},
+		XXX_unrecognized:     []byte{},
+		XXX_sizecache:        0,
+	}
+
+	return timeSeries{blockSample.lblsHash, ts}
+}
+
+func NewPromLoader(dataDir string) (PromLoader, error) {
+	db, err := tsdb.OpenDBReadOnly(dataDir, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error starting Prometheus TSDB in read-only: %v", err)
+	}
+	blocks, err := db.Blocks()
+	if err != nil {
+		return nil, fmt.Errorf("error loading data blocks: %v", err)
+	}
+	return &promLoader{db: db, blocks: blocks}, nil
+}
+
+func (loader *promLoader) Iterator() PromIterator {
+	return &promIterator{blocks: loader.blocks, curSampleIdx: -1, curBlockIdx: -1}
+}
+
+func (loader *promLoader) Close() error {
+	return loader.db.Close()
+}
+
+type IngestFunc func(context.Context, *prompb.WriteRequest) (uint64, uint64, error)
+
+type sampleIngestor struct {
+	samples   PromIterator
+	shards    []chan prompb.TimeSeries
+	batchSize int
+	rate      *rate.Limiter
+}
+
+// NewSampleIngestor creates sampleIngestor responsible for splitting samples into shards and sending them down the ingestion path
+// zero sampleRate means no limits
+func NewSampleIngestor(shards int, samples PromIterator, batchSize int, sampleRate int) *sampleIngestor {
+	var r *rate.Limiter
+	if sampleRate != 0 {
```

Comment:
We could use `rate.Inf` for `sampleRate == 0`. Would simplify the rate limiting logic (no need for checking nils).

Reply:
Have you benchmarked this? :grin: